src/System.Private.CoreLib/shared/System/Marvin.cs

```diff
@@ -28,73 +28,194 @@ internal static partial class Marvin
-                    Block(ref p0, ref p1);
-                    goto case 1;
-
-                case 1:
-                    p0 += 0x8000u | Unsafe.AddByteOffset(ref data, byteOffset);
-                    break;
-
-                case 6:
-                    p0 += Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref data, byteOffset));
-                    byteOffset += 4;
-                    Block(ref p0, ref p1);
-                    goto case 2;
-
-                case 2:
-                    p0 += 0x800000u | Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref data, byteOffset));
-                    break;
-
-                case 7:
-                    p0 += Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref data, byteOffset));
-                    byteOffset += 4;
-                    Block(ref p0, ref p1);
-                    goto case 3;
-
-                case 3:
-                    p0 += 0x80000000u | (((uint)(Unsafe.AddByteOffset(ref data, byteOffset + 2))) << 16) | (uint)(Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref data, byteOffset)));
-                    break;
-
-                default:
-                    Debug.Fail("Should not get here.");
-                    break;
+                partialResult >>= 8; // make some room for the 0x80 byte
+                partialResult |= 0x80000000u; // put the 0x80 byte at the beginning
+                partialResult >>= (~count << 3) & 0x1F; // shift out all previously consumed bytes
             }
+            else
+            {
+                partialResult <<= 8; // make some room for the 0x80 byte
+                partialResult |= 0x80u; // put the 0x80 byte at the end
+                partialResult <<= (~count << 3) & 0x1F; // shift out all previously consumed bytes
+            }
+
+        DoFinalRoundsAndReturn:
 
+            // Now that we've computed the final partial result, merge it in and run two rounds of
+            // the block function to finish out the Marvin algorithm.
+
+            p0 += partialResult;
             Block(ref p0, ref p1);
             Block(ref p0, ref p1);
 
             return (int)(p1 ^ p0);
+
+        InputTooSmallToEnterMainLoop:
+
+            // We can't run the main loop, but we might still have 4 or more bytes available to us.
+            // If so, jump to the 4 .. 7 bytes logic immediately after the main loop.
+
+            if ((count & 4) != 0)
+            {
+                goto Between4And7BytesRemain;
+            }
+
+            // We had only 0 .. 3 bytes to begin with, so we can't perform any 32-bit reads.
+            // This means that we're going to be building up the final result right away and
+            // will only ever run two rounds total of the block function. Let's initialize
+            // the partial result to "no data".
+
+            if (BitConverter.IsLittleEndian)
+            {
+                partialResult = 0x80u;
+            }
+            else
```

Comment:
Is there a way to test big endian? (Outside of wait for Mono to try it)

Reply:
I don't have a way to do so. The best bet right now IMO is to create more unit tests and make sure that those tests are running on big-endian platforms. FWIW the existing code (before any of these changes) doesn't output correct values on big-endian anyway, so I hope I didn't make the problem worse. :)

Reply:
/cc @NattyNarwhal

Reply:
I can't compile this with mcs; it doesn't like `data = ref Unsafe.AddByteOffset(ref data, 8);`.

```
../../../external/corefx/src/Common/src/CoreLib/System/Marvin.cs(71,23): error CS1525: Unexpected symbol `ref'
../../../external/corefx/src/Common/src/CoreLib/System/Marvin.cs(71,60): error CS1525: Unexpected symbol `;', expecting `(', `.', `[', or `<operator>'
```

Mono hasn't hasn't bumped Roslyn AFAIK, to the version that supports BE. (if it doesn't just reveal some glaring issue itself) The CoreFX is older, but the file at `src/Common/src/CoreLib/System/Marvin.cs` is basically the thing bar an ifdef for Mono that I patched into this one.